Automattic/wp-calypso
# Checkout: refactor `getCheckoutCompleteRedirectPath` to clarify usage and add tests

#### Changes proposed in this Pull Request

This PR refactors the `getCheckoutCompleteRedirectPath` method of the `Checkout` component. It had some complex logic which was hard to reason about; this refactor should not change any behavior but hopefully makes it easier to read and understand.

A large part of this diff is moving methods from inside the component to pure functions outside. This should help make argument dependencies more clear and also makes the functions more easily testable.

This was a needed refactor anyway, but its main purpose is to clarify the url-choosing behavior so we can replicate it in #38725. 

**NOTE: Because of the size of this diff, which is mostly adding tests, it may be easiest to review the commits one-at-a-time. I've made an effort to keep each commit as small as possible.**

#### Testing instructions

This is difficult to test because of all the different possible conditions involved and because of how much work it takes to test each one. Therefore it may be necessary to rely on the automated tests to verify the behavior, but it's worth trying at least one purchase flow to be sure that it works. To do this, first test on master (or production) and then test on this branch: add an item to your cart, complete the purchase (you can sandbox the store to make this a test purchase), and be sure that you end up at the same thank-you page either way.

This PR adds 37 unit tests which cover all the code paths of the function. You can verify this by doing the following:

- Check out the commit after the tests are added but before the refactor: `git checkout c263979c6f`
- Run the tests: `npx jest --verbose -c test/client/jest.config.js client/my-sites/checkout/checkout/test/checkout.js`
